File: Python3/3128_right_triangles.py
# https://leetcode.cn/problems/right-triangles/

class Solution:
    def numberOfRightTriangles(self, grid):
        """
        Inputs:
            grid: list[list[int]]
        Outputs:
            res: int
        """
        rows, cols = len(grid), len(grid[0])
        cnts = [[[0, 0, 0, 0] for _ in range(cols)] for _ in range(rows)]
        for i in range(rows):
            # Left.
            cnt = grid[i][0]
            for j in range(1, cols):
                cnts[i][j][0] = cnt
                cnt += grid[i][j]
            # Right.
            cnt = grid[i][cols-1]
            for j in range(cols-2, -1, -1):
                cnts[i][j][1] = cnt
                cnt += grid[i][j]
        
        for j in range(cols):
            # Above.
            cnt = grid[0][j]
            for i in range(1, rows):
                cnts[i][j][2] = cnt
                cnt += grid[i][j]
            # Below.
            cnt = grid[rows-1][j]
            for i in range(rows-2, -1, -1):
                cnts[i][j][3] = cnt
                cnt += grid[i][j]
        
        res = 0
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1:
                    res += (cnts[i][j][0] + cnts[i][j][1]) * (cnts[i][j][2] + cnts[i][j][3])
        return res

File: Cpp/3128_right_triangles.cpp
// https://leetcode.cn/problems/right-triangles/

class Solution {
public:
    long long numberOfRightTriangles(vector<vector<int>>& grid) {
        int rows = grid.size(), cols = grid[0].size();
        vector<vector<vector<int>>> cnts(rows, vector<vector<int>>(cols, vector<int>(4, 0)));
        for (int i = 0; i < rows; ++i) {
            // Left.
            int cnt = grid[i][0];
            for (int j = 1; j < cols; ++j) {
                cnts[i][j][0] = cnt;
                cnt += grid[i][j];
            }
            // Right.
            cnt = grid[i][cols-1];
            for (int j = cols-2; j >= 0; --j) {
                cnts[i][j][1] = cnt;
                cnt += grid[i][j];
            }
        }
        for (int j = 0; j < cols; ++j) {
            // Above.
            int cnt = grid[0][j];
            for (int i = 1; i < rows; ++i) {
                cnts[i][j][2] = cnt;
                cnt += grid[i][j];
            }
            // Below.
            cnt = grid[rows-1][j];
            for (int i = rows-2; i >= 0; --i) {
                cnts[i][j][3] = cnt;
                cnt += grid[i][j];
            }
        }
        long long res = 0;
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (grid[i][j] == 1) {
                    res += ((long long)cnts[i][j][0] + cnts[i][j][1])
                        * ((long long)cnts[i][j][2] + cnts[i][j][3]);
                }
            }
        }
        return res;
    }
};
